models/user.py

```diff
@@ -104,6 +107,10 @@ def json(self):
             "created_at": Time.format_date(self.created_at),
             "updated_at": Time.format_date(self.updated_at),
         }
+        return {**base_json, **self.serialize()}
+
+    def serialize(self):
+        return {}
```

Comment:
Not a critique, but a question. Why does the `serialize` method return an empty dictionary? Is this just an abstract method for subclasses to implement so that the `json` method itself doesn't have to be overwritten by each subclass?

Reply:
Only critiques accepted. No questions allowed 😉 

Thanks for asking @yunnie You are spot on, it kinda acts as an abstract method. This method is optional for our subclasses to implement. I've heard this referred to as a hook method. This allows the subclasses to hook into the json method without needing to know how it works. The other option as you mentioned is to have the subclass override the json method. Perhaps calling super and then needing to know how to take those results and combine it with the extra functionality the subclass wants to provide. That tends to lead to a lot of duplicate code and knowledge in the subclasses on how to interact with the json method. It is a better coding style to use hook methods instead. There are other reasons too.

tl/dr
it returns an empty hash because the json method still needs to function without subclasses. If I wanted to make serialize required then I've done this in the past.

```python
def serialize(self):
    raise NotImplementedError('Some message for devs')
```

FYI: Pending users will be in our system, and when they're serialized then they will be of type `UserModel` (since they haven't been assigned a user type/role yet), and so this serialize method will be called and the empty hash will be returned.